### PR TITLE
DOCS-#2940: add documentation for asv 

### LINKED ISSUE #2940: Document asv bench and CI
This should document what we do in asv benchmarking, how we produce the results (and steps for any 3rd party to reproduce the measurements).

It should also document everything about how we run CI on GitHub and TeamCity

### PR DESCRIPTION
**Pull request by anmyachev**
Signed-off-by: Anatoly Myachev <anatoly.myachev@intel.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #2940 
- [x] tests added and passing

### CODE REVIEW

File: asv_bench/benchmarks/scalability/__init__.py [hunk 1]
```diff
@@ -10,3 +10,5 @@
 # the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
 # ANY KIND, either express or implied. See the License for the specific language
 # governing permissions and limitations under the License.
+
+"""Modin benchmarks with MODIN_CPUS change"""
```
Comment (vnlitvinov) [inline asv_bench/benchmarks/scalability/__init__.py:14]:
```suggestion
"""Benchmarks measuring how Modin performance scales when MODIN_CPUS are changed"""
```

File: asv_bench/benchmarks/utils.py [hunk 2]
```diff
@@ -11,12 +11,20 @@
 # ANY KIND, either express or implied. See the License for the specific language
 # governing permissions and limitations under the License.
 
+"""
+This module contains the functionality that is used when benchmarking modin commits.
+In the case of using utilities from the main modin code, there is a chance that when
+benchmarking old commits, the utilities changed, which in turn can unexpectedly affect
+the performance results.
```
Comment (vnlitvinov) [inline asv_bench/benchmarks/utils.py:18]:
```suggestion
the performance results, hence some utility functions are duplicated here.
```

File: asv_bench/benchmarks/utils.py [hunk 3]
```diff
@@ -107,7 +127,26 @@ class weakdict(dict):
 dataframes_cache = dict()
 
 
-def gen_int_data(nrows, ncols, rand_low, rand_high):
+def gen_int_data(nrows: int, ncols: int, rand_low: int, rand_high: int) -> dict:
+    """
+    Generate int data with caching.
```
Comment (vnlitvinov) [inline asv_bench/benchmarks/utils.py:132]: what does "caching" mean here?

File: asv_bench/benchmarks/utils.py [hunk 4]
```diff
@@ -211,20 +294,57 @@ def generate_dataframe(
     return df
 
 
-def random_string():
-    return str(uuid.uuid1())
+def random_string() -> str:
+    """
+    Create 36 character random strings using functionality from `uuid` module.
```
Comment (vnlitvinov) [inline asv_bench/benchmarks/utils.py:299]:
```suggestion
    Create 36-character random strings.
```
Docstring does not have to explain _how_ stuff is done unless it is critical for the function's role.

Based on function name I think we don't care that these numbers are actually UUID-s. If that isn't true (i.e. if we rely somewhere on them being UUIDs) we should rename the function.

File: asv_bench/benchmarks/utils.py [hunk 5]
```diff
@@ -211,20 +294,57 @@ def generate_dataframe(
...
+    Return
+    ------
+    str
+    """
+    return str(uuid.uuid4())
+
+
+def random_columns(df_columns: list, columns_number: int) -> list:
+    """
+    Create list of random columns from a given sequence.
```
Comment (vnlitvinov) [inline asv_bench/benchmarks/utils.py:310]:
```suggestion
    Pick sublist of random columns from a given sequence.
```

File: asv_bench/benchmarks/utils.py [hunk 6]
```diff
@@ -211,20 +294,57 @@ def generate_dataframe(
...
 
+    Parameters
+    ----------
+    df_columns: list
+    columns_number: int
```
Comment (vnlitvinov) [inline asv_bench/benchmarks/utils.py:315]: add here something like "how many columns to pick"

File: asv_bench/benchmarks/utils.py [hunk 7]
```diff
@@ -211,20 +294,57 @@ def generate_dataframe(
...
+    list
+    """
     return list(random_state.choice([True, False], size=number))
 
 
-def execute(df):
-    "Make sure the calculations are done."
+def execute(df: Union[pd.DataFrame, pandas.DataFrame]):
+    """
+    Make sure the calculations are done.
```
Comment (vnlitvinov) [inline asv_bench/benchmarks/utils.py:342]:
```suggestion
    Make sure the calculations are finished.
```

**Review by vnlitvinov — COMMENTED** [review]
Note that you can batch-apply suggestions from "Files changed" pane, so there would be a single commit of all of them.

File: asv_bench/README.md [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+the hash in the corresponding result files.
+
+## Pipeline for displaying results in a dashboard
+
+Step 1: checking benchmarks for validity, runs in PRs CI.
+  During the test, the benchmarks are run once on small data.
+  The implementation can be found in `test-asv-benchmarks` job of [ci.yml](https://github.com/modin-project/modin/blob/master/.github/workflows/ci.yml)
+
+Step 2: running benchmarks with saving the results in [modin-bench@master](https://github.com/modin-project/modin-bench).
+  The launch takes place on our own server using specific TeamCity configuration.
```
Comment (Garra1980) [inline asv_bench/README.md:66]: "Our own" sounds a bit ambiguous to me. Not sure it is clear in public context

File: asv_bench/benchmarks/utils.py [hunk 9]
```diff
@@ -90,7 +98,19 @@
...
 
 
-def translator_groupby_ngroups(groupby_ngroups, shape):
+def translator_groupby_ngroups(groupby_ngroups: Union[str, int], shape: tuple) -> int:
+    """
+    Translate a string representation of the number of groups, into a number.
+
+    Parameters
+    ----------
+    groupby_ngroups:
```
Comment (Garra1980) [inline asv_bench/benchmarks/utils.py:107]: Some description for params is needed?

File: asv_bench/README.md [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+  This step starts as scheduled (now every half hour), subject to the presence of new commits in the modin master branch.
+  Command to run benchmarks: `asv run HASHFILE:hashfile.txt --show-stderr --machine xeon-e5 --launch-method=spawn`.
+  In the file `hashfile.txt` is the last modin commit hash.
+  Writing to a `modin-bench@master` triggers 3 step of the pipeline.
+
+3 step: converting the results to html representation, which is saved in [modin-bench@gh-pages](https://github.com/modin-project/modin-bench)
+  The implementation can be found in `deploy-gh-pages` job of [push.yml](https://github.com/modin-project/modin-bench/blob/master/.github/workflows/push.yml)
+
+Basic actions for 2 step:
+* install packages to proper installation ray dependencies via pip:
```
[the remark this answers is not in the document]
Comment (anmyachev) [reply asv_bench/README.md]: Now this is not needed, apparently it has been preserved from fairly old dependencies. So removed these lines.

**Comment by anmyachev** [comment]
@vnlitvinov @Garra1980 ready for review

File: asv_bench/README.md [hunk 11]
```diff
@@ -2,86 +2,85 @@
...
 
   ```
   asv run [start_hash]..[end_hash] --launch-method=spawn
   asv publish
   asv preview
   ```
 
-For more consistent results, you may need to use the following parameters:
+For more consistent results, you may need to use the following parameters whose
+[description is in](https://asv.readthedocs.io/en/stable/benchmarks.html?highlight=sample_time#timing-benchmarks):
```
Comment (vnlitvinov) [inline asv_bench/README.md:26]:
```suggestion
For more consistent results, you may need to use the following parameters which
description is in [ASV docs](https://asv.readthedocs.io/en/stable/benchmarks.html?highlight=sample_time#timing-benchmarks):
```

File: asv_bench/README.md [hunk 12]
```diff
@@ -2,86 +2,85 @@
...
+  This step starts as scheduled (now every half hour), subject to the presence of new commits in the Modin `master` branch.
   Command to run benchmarks: `asv run HASHFILE:hashfile.txt --show-stderr --machine xeon-e5 --launch-method=spawn`.
   In the file `hashfile.txt` is the last modin commit hash.
   Writing to a `modin-bench@master` triggers 3 step of the pipeline.
 
-3 step: converting the results to html representation, which is saved in [modin-bench@gh-pages](https://github.com/modin-project/modin-bench)
+Step 3: converting the results to html representation, which is saved in [modin-bench@gh-pages](https://github.com/modin-project/modin-bench)
   The implementation can be found in `deploy-gh-pages` job of [push.yml](https://github.com/modin-project/modin-bench/blob/master/.github/workflows/push.yml)
 
 Basic actions for 2 step:
```
Comment (vnlitvinov) [inline asv_bench/README.md:78]:
```suggestion
Basic actions for step 2:
```

File: asv_bench/README.md [hunk 13]
```diff
@@ -2,86 +2,85 @@
...
-* install packages to proper installation ray dependencies via pip:
-  * gcc
-  * python3-dev
 * setup environment variable:
   * export MODIN_TEST_DATASET=Big
   * export MODIN_CPUS=44
 * setup git client
 * prepare json file with machine description
```
Comment (vnlitvinov) [inline asv_bench/README.md:83]: this could use a more elaborate documentation - how do I prepare a file?

File: asv_bench/benchmarks/utils.py [hunk 14]
```diff
@@ -236,33 +249,67 @@ def gen_data(
...
+        number of columns for which `groupby` will be called in the future;
+        to get more stable performance results, we need to have the same number of values
+        in each group every benchmarking time
+    count_groups: int or None
+        count of groups in groupby columns
+
+    Return
+    ------
+    modin.DataFrame or pandas.DataFrame [and list of groupby columns names if
+        columns for groupby were be generated]
```
Comment (vnlitvinov) [inline asv_bench/benchmarks/utils.py:294]:
```suggestion
    modin.DataFrame or pandas.DataFrame [and list of groupby columns names if
        columns for groupby were generated]
```

File: asv_bench/benchmarks/utils.py [hunk 15]
```diff
@@ -296,7 +343,7 @@ def generate_dataframe(
 
 def random_string() -> str:
     """
-    Create 36 character random strings using functionality from `uuid` module.
+    Create 36-character random strings.
```
Comment (vnlitvinov) [inline asv_bench/benchmarks/utils.py:346]:
```suggestion
    Create a 36-character random string.
```

**Review by vnlitvinov — COMMENTED** [review]
Mostly there

File: asv_bench/README.md [hunk 16]
```diff
@@ -75,12 +75,14 @@ Step 2: running benchmarks with saving the results in [modin-bench@master](https
...
 
-Basic actions for 2 step:
+Basic actions for step 2:
 * setup environment variable:
   * export MODIN_TEST_DATASET=Big
   * export MODIN_CPUS=44
 * setup git client
 * prepare json file with machine description
+  * This file should be placed in the user's home directory.
+  * [Example](https://github.com/modin-project/modin-bench/blob/master/results/xeon-e5/machine.json)
```
Comment (vnlitvinov) [inline asv_bench/README.md:85]: Example is a good thing, but link to docs describing it would be an awesome addition as well :)

**Review by vnlitvinov — APPROVED** [review]

File: asv_bench/README.md (hunk 16 above)
Comment (anmyachev) [reply asv_bench/README.md]: link is added.

File: asv_bench/README.md [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+  The implementation can be found in `deploy-gh-pages` job of [push.yml](https://github.com/modin-project/modin-bench/blob/master/.github/workflows/push.yml)
+
+Basic actions for step 2:
+* setup environment variable:
+  * export MODIN_TEST_DATASET=Big
+  * export MODIN_CPUS=44
+* setup git client
+* prepare json file with machine description
+  * This file should be placed in the user's home directory.
+  * Automatically ASV does not always correctly create the file with the description of the machine.
```
Comment (vnlitvinov) [inline asv_bench/README.md:85]:
```suggestion
  * ASV does not always automatically create the file with the description of the machine correctly (e.g. due to being run in a container).
```

**Review by vnlitvinov — DISMISSED** [review]

**Review by vnlitvinov — APPROVED** [review]
LGTM

**Review by Garra1980 — APPROVED** [review]

**Comment by anmyachev** [comment]
@devin-petersohn please review

**Review by devin-petersohn — APPROVED** [review]
Thanks @anmyachev, looks great!

File: asv_bench/benchmarks/utils.py [hunk 18]
```diff
@@ -90,7 +98,20 @@
...
 
 
-def translator_groupby_ngroups(groupby_ngroups, shape):
+def translator_groupby_ngroups(groupby_ngroups: Union[str, int], shape: tuple) -> int:
+    """
+    Translate a string representation of the number of groups, into a number.
+
+    Parameters
+    ----------
+    groupby_ngroups: str or int
```
Comment (YarShev) [inline asv_bench/benchmarks/utils.py:107]: I prefer using indent between a parameter name and colon as it is presented in numpydoc guide :)

Comment (devin-petersohn) [reply asv_bench/benchmarks/utils.py:107]: @YarShev it should be a quick fix.

Do our doc tests not properly check the numpydoc standard?

Comment (YarShev) [reply asv_bench/benchmarks/utils.py:107]: It seems that `pydocstyle --convention=numpy` doesn't check it. Although, the guide says the following: `The colon must be preceded by a space, or omitted if the type is absent.`

Comment (Garra1980) [reply asv_bench/benchmarks/utils.py:107]: Should be "groupby_ngroups : str or int" ie space before :

File: asv_bench/benchmarks/utils.py [hunk 19]
```diff
@@ -90,7 +98,20 @@
...
+        number of groups that will be used in `groupby` operation
+    shape: tuple
+
+    Return
```
Comment (Garra1980) [inline asv_bench/benchmarks/utils.py:111]: Should be "Returns" ie with s at the end

**Pull request merged**